### PR TITLE
core/unit: make unit dependency always bidirectional

### LINKED ISSUE #27742: Heap-use-after-free if we fail to add a reverse dependency to a unit
### systemd version the issue has been seen with

latest main

### Used distribution

Fedora 37

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

In my ongoing adventures with Nallocfuzz, it managed to trigger one heap-use-after-free:

```
==7==ERROR: AddressSanitizer: heap-use-after-free on address 0x619000003798 at pc 0x7f18c9b24ac6 bp 0x7ffd69d4cb10 sp 0x7ffd69d4cb08
READ of size 8 at 0x619000003798 thread T0
SCARINESS: 51 (8-byte-read-heap-use-after-free)
    #0 0x7f18c9b24ac5 in unit_clear_dependencies /work/build/../../src/systemd/src/core/unit.c:640:58
    #1 0x7f18c9b24ac5 in unit_free /work/build/../../src/systemd/src/core/unit.c:786:9
    #2 0x7f18c9a0668b in manager_clear_jobs_and_units /work/build/../../src/systemd/src/core/manager.c:1591:17
    #3 0x7f18c9a059d5 in manager_free /work/build/../../src/systemd/src/core/manager.c:1625:9
    #4 0x4de361 in manager_freep /work/build/../../src/systemd/src/core/manager.h:505:1
    #5 0x4de361 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-manager-serialize.c:40:1
    #6 0x4de858 in NaloFuzzerTestOneInput (/build/fuzz-manager-serialize+0x4de858)
    #7 0x4fd7e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #8 0x4fe305 in fuzzer::Fuzzer::TryDetectingAMemoryLeak(unsigned char const*, unsigned long, bool) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:687:3
    #9 0x4e8f8e in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:329:8
    #10 0x4ee7ec in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #11 0x4eef98 in LLVMFuzzerRunDriver /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:925:10
    #12 0x4deab5 in main (/build/fuzz-manager-serialize+0x4deab5)
    #13 0x7f18c8432082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #14 0x41f81d in _start (/build/fuzz-manager-serialize+0x41f81d)

DEDUP_TOKEN: unit_clear_dependencies--unit_free--manager_clear_jobs_and_units
0x619000003798 is located 24 bytes inside of 1104-byte region [0x619000003780,0x619000003bd0)
freed by thread T0 here:
    #0 0x4a09b2 in __interceptor_free /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:52:3
    #1 0x7f18c9b244a3 in unit_free /work/build/../../src/systemd/src/core/unit.c:885:16
    #2 0x7f18c9a0668b in manager_clear_jobs_and_units /work/build/../../src/systemd/src/core/manager.c:1591:17
    #3 0x7f18c9a059d5 in manager_free /work/build/../../src/systemd/src/core/manager.c:1625:9
    #4 0x4de361 in manager_freep /work/build/../../src/systemd/src/core/manager.h:505:1
    #5 0x4de361 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-manager-serialize.c:40:1
    #6 0x4de858 in NaloFuzzerTestOneInput (/build/fuzz-manager-serialize+0x4de858)
    #7 0x4fd7e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #8 0x4fe305 in fuzzer::Fuzzer::TryDetectingAMemoryLeak(unsigned char const*, unsigned long, bool) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:687:3
    #9 0x4e8f8e in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:329:8
    #10 0x4ee7ec in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #11 0x4eef98 in LLVMFuzzerRunDriver /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:925:10
    #12 0x4deab5 in main (/build/fuzz-manager-serialize+0x4deab5)
    #13 0x7f18c8432082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

DEDUP_TOKEN: __interceptor_free--unit_free--manager_clear_jobs_and_units
previously allocated by thread T0 here:
    #0 0x4a0e2e in __interceptor_calloc /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:77:3
    #1 0x4de4f1 in calloc (/build/fuzz-manager-serialize+0x4de4f1)
    #2 0x7f18c9b19a57 in unit_new /work/build/../../src/systemd/src/core/unit.c:98:13
    #3 0x7f18c9a0dabf in manager_load_unit_prepare /work/build/../../src/systemd/src/core/manager.c:2304:39
    #4 0x7f18c9a0c26f in manager_load_unit /work/build/../../src/systemd/src/core/manager.c:2344:13
    #5 0x7f18c9b2bfcc in unit_add_two_dependencies_by_name /work/build/../../src/systemd/src/core/unit.c:3331:13
    #6 0x7f18c9aa9974 in slice_add_default_dependencies /work/build/../../src/systemd/src/core/slice.c:76:13
    #7 0x7f18c9aa9974 in slice_load /work/build/../../src/systemd/src/core/slice.c:182:13
    #8 0x7f18c9b2fa8d in unit_load /work/build/../../src/systemd/src/core/unit.c:1707:13
    #9 0x7f18c9a0cc13 in manager_dispatch_load_queue /work/build/../../src/systemd/src/core/manager.c:2216:17
    #10 0x7f18c9a0c28f in manager_load_unit /work/build/../../src/systemd/src/core/manager.c:2349:9
    #11 0x7f18c9a008df in manager_deserialize_one_unit /work/build/../../src/systemd/src/core/manager-serialize.c:221:13
    #12 0x7f18c9a008df in manager_deserialize_units /work/build/../../src/systemd/src/core/manager-serialize.c:253:21
    #13 0x7f18c9a008df in manager_deserialize /work/build/../../src/systemd/src/core/manager-serialize.c:598:16
    #14 0x4de318 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-manager-serialize.c:37:16
    #15 0x4de858 in NaloFuzzerTestOneInput (/build/fuzz-manager-serialize+0x4de858)
    #16 0x4fd7e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #17 0x4fe305 in fuzzer::Fuzzer::TryDetectingAMemoryLeak(unsigned char const*, unsigned long, bool) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:687:3
    #18 0x4e8f8e in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:329:8
    #19 0x4ee7ec in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #20 0x4eef98 in LLVMFuzzerRunDriver /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:925:10
    #21 0x4deab5 in main (/build/fuzz-manager-serialize+0x4deab5)
    #22 0x7f18c8432082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

DEDUP_TOKEN: __interceptor_calloc--calloc--unit_new
SUMMARY: AddressSanitizer: heap-use-after-free /work/build/../../src/systemd/src/core/unit.c:640:58 in unit_clear_dependencies
Shadow bytes around the buggy address:
  0x0c327fff86a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff86b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff86c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff86d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff86e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c327fff86f0: fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff8700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff8710: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff8720: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff8730: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c327fff8740: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
```

After some debugging (in a naive hope that I could fix it) it seems like this happens when we fail to add an inverse dependency to a unit, specifically in this part:

https://github.com/systemd/systemd/blob/139b6fb16f9a3699db55a4eb66426df165a4bcf8/src/core/unit.c#L3196-L3201

The Nallocfuzz fail can be easily reproduced by simply monkey-patching the code:
```diff
diff --git a/src/core/unit.c b/src/core/unit.c
index 3393138bac..4a86dec423 100644
--- a/src/core/unit.c
+++ b/src/core/unit.c
@@ -3194,6 +3194,7 @@ int unit_add_dependency(
         notify = r > 0;
 
         if (inverse_table[d] != _UNIT_DEPENDENCY_INVALID && inverse_table[d] != d) {
+                return -EINVAL;
                 r = unit_add_dependency_hashmap(&other->dependencies, inverse_table[d], u, 0, mask);
                 if (r < 0)
                         return r;
```

building with ASan, and running `fuzz-manager-serialize`:
```
$ build-san/fuzz-manager-serialize test/fuzz/fuzz-manager-serialize/state
test/fuzz/fuzz-manager-serialize/state... =================================================================
==1491138==ERROR: AddressSanitizer: heap-use-after-free on address 0x61b000000eb0 at pc 0x7f9088d8c732 bp 0x7fffc3734ae0 sp 0x7fffc3734ad8
READ of size 8 at 0x61b000000eb0 thread T0
SCARINESS: 51 (8-byte-read-heap-use-after-free)
    #0 0x7f9088d8c731 in unit_clear_dependencies ../src/core/unit.c:635
    #1 0x7f9088d8ee8c in unit_free ../src/core/unit.c:779
    #2 0x7f9088b6268c in manager_clear_jobs_and_units ../src/core/manager.c:1590
    #3 0x7f9088b635e2 in manager_free ../src/core/manager.c:1623
...
```

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
This also fixes several issues for `JoinsNamespaceOf=`.
Fixes #27742.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
Thanks! I rebuild the fuzzers and now I keep hitting a different issue:

```
manager_deserialize_one_unit: deserializing stem.slice
=================================================================
==7==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x621000005a10 in thread T0
SCARINESS: 40 (bad-free)
    #0 0x4a09b2 in __interceptor_free /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:52:3
    #1 0x7f5bfdd65a0d in freep /work/build/../../src/systemd/src/basic/alloc-util.h:107:22
    #2 0x7f5bfdd65a0d in unit_get_dependency_hashmap_per_type /work/build/../../src/systemd/src/core/unit.c:3067:9
    #3 0x7f5bfdd46eb3 in unit_add_dependency_impl /work/build/../../src/systemd/src/core/unit.c:3136:22
    #4 0x7f5bfdd2d6c2 in unit_add_dependency /work/build/../../src/systemd/src/core/unit.c:3250:13
    #5 0x7f5bfdd2fc5b in unit_add_two_dependencies /work/build/../../src/systemd/src/core/unit.c:3275:13
    #6 0x7f5bfdd2fc5b in unit_add_slice_dependencies /work/build/../../src/systemd/src/core/unit.c:1507:24
    #7 0x7f5bfdd2fc5b in unit_load /work/build/../../src/systemd/src/core/unit.c:1668:21
    #8 0x7f5bfdc0cb93 in manager_dispatch_load_queue /work/build/../../src/systemd/src/core/manager.c:2214:17
    #9 0x7f5bfdc0c20f in manager_load_unit /work/build/../../src/systemd/src/core/manager.c:2345:9
    #10 0x7f5bfdc008df in manager_deserialize_one_unit /work/build/../../src/systemd/src/core/manager-serialize.c:221:13
    #11 0x7f5bfdc008df in manager_deserialize_units /work/build/../../src/systemd/src/core/manager-serialize.c:253:21
    #12 0x7f5bfdc008df in manager_deserialize /work/build/../../src/systemd/src/core/manager-serialize.c:598:16
    #13 0x4de318 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-manager-serialize.c:37:16
    #14 0x4de858 in NaloFuzzerTestOneInput (/build/fuzz-manager-serialize+0x4de858)
    #15 0x4fd7e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #16 0x4fcfca in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #17 0x4fee34 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
    #18 0x4ff069 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #19 0x4ee6cf in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #20 0x4eef98 in LLVMFuzzerRunDriver /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:925:10
    #21 0x4deab5 in main (/build/fuzz-manager-serialize+0x4deab5)
    #22 0x7f5bfc632082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #23 0x41f81d in _start (/build/fuzz-manager-serialize+0x41f81d)

DEDUP_TOKEN: __interceptor_free--freep--unit_get_dependency_hashmap_per_type
0x621000005a10 is located 2320 bytes inside of 4096-byte region [0x621000005100,0x621000006100)
allocated by thread T0 here:
    #0 0x4a0c56 in __interceptor_malloc /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:69:3
    #1 0x4de669 in malloc (/build/fuzz-manager-serialize+0x4de669)
    #2 0x7f5bfcf54469 in mempool_alloc_tile /work/build/../../src/systemd/src/basic/mempool.c:49:21
    #3 0x7f5bfcf54708 in mempool_alloc0_tile /work/build/../../src/systemd/src/basic/mempool.c:68:13
    #4 0x7f5bfcf1c695 in hashmap_base_new /work/build/../../src/systemd/src/basic/hashmap.c:785:24
    #5 0x7f5bfcf1ca4f in hashmap_base_ensure_allocated /work/build/../../src/systemd/src/basic/hashmap.c:836:13
    #6 0x7f5bfcf1ca4f in _hashmap_ensure_allocated /work/build/../../src/systemd/src/basic/hashmap.c:845:16
    #7 0x7f5bfdc03093 in manager_new /work/build/../../src/systemd/src/core/manager.c:952:13
    #8 0x4de1dc in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-manager-serialize.c:26:15
    #9 0x4de858 in NaloFuzzerTestOneInput (/build/fuzz-manager-serialize+0x4de858)
    #10 0x4fd7e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #11 0x4feb94 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:804:3
    #12 0x4ff069 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #13 0x4ee6cf in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #14 0x4eef98 in LLVMFuzzerRunDriver /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:925:10
    #15 0x4deab5 in main (/build/fuzz-manager-serialize+0x4deab5)
    #16 0x7f5bfc632082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

DEDUP_TOKEN: __interceptor_malloc--malloc--mempool_alloc_tile
SUMMARY: AddressSanitizer: bad-free /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:52:3 in __interceptor_free
```

**Comment by mrc0mmand** [comment]
Yup, now the fuzzer seems to be happy, or at least it hasn't crashed in the past ~40 minutes, nice!

File: src/core/unit.c [hunk 1]
```diff
@@ -3120,7 +3120,8 @@ int unit_add_dependency(
                 [UNIT_RELOAD_PROPAGATED_FROM] = UNIT_PROPAGATES_RELOAD_TO,
                 [UNIT_PROPAGATES_STOP_TO]     = UNIT_STOP_PROPAGATED_FROM,
                 [UNIT_STOP_PROPAGATED_FROM]   = UNIT_PROPAGATES_STOP_TO,
-                [UNIT_JOINS_NAMESPACE_OF]     = UNIT_JOINS_NAMESPACE_OF, /* symmetric! 👓 */
+                [UNIT_JOINS_NAMESPACE_OF]     = UNIT_JOINED_NAMESPACE_BY,
```
Comment (dtardon) [inline src/core/unit.c]: Maybe just use `UNIT_REFERENCED_BY` as the inverse dep? That should be enough for the purpose, shouldn't it?

**Comment by yuwata** [comment]
Now, slightly different approach is used for `JoinsNamespaceOf=`. PTAL.

File: src/core/unit.c [hunk 2]
```diff
@@ -4890,7 +4889,12 @@ int unit_setup_exec_runtime(Unit *u) {
         assert(ec);
 
         /* Try to get it from somebody else */
-        UNIT_FOREACH_DEPENDENCY(other, u, UNIT_ATOM_JOINS_NAMESPACE_OF) {
+        UnitDependencyInfo info;
+        UNIT_FOREACH_DEPENDENCY_WITH_INFO(other, info, u, UNIT_ATOM_JOINS_NAMESPACE_OF) {
+                /* UNIT_JOINS_NAMESPACE_OF is symmetric. */
+                if (info.origin_mask == 0)
+                        continue;
```
Comment (poettering) [inline src/core/unit.c]: hmm, i don't understand check? can you elaborate?

File: src/core/unit.c [hunk 3]
```diff
@@ -3126,12 +3113,84 @@ int unit_add_dependency(
...
+        /* Update dependency info. */
+        u_info.origin_mask |= mask;
+        other_info.destination_mask |= mask;
+
+        /* Save updated dependency info. */
+        if (u_info.data != u_info_old.data) {
+                if (u_info_old.data)
+                        r = hashmap_update(u_deps, other, u_info.data);
+                else
+                        r = hashmap_put(u_deps, other, u_info.data);
```
Comment (poettering) [inline src/core/unit.c]: just use hashmap_replace()?

File: src/core/unit.c [hunk 4]
```diff
@@ -3126,12 +3113,84 @@ int unit_add_dependency(
...
+                        return r;
+
+                flags = NOTIFY_DEPENDENCY_UPDATE_FROM;
+        }
+
+        if (other_info.data != other_info_old.data) {
+                if (other_info_old.data)
+                        r = hashmap_update(other_deps, u, other_info.data);
+                else
+                        r = hashmap_put(other_deps, u, other_info.data);
```
Comment (poettering) [inline src/core/unit.c]: ditto

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
looks good to me, though i don't grok that one check you added, see above

**Comment by mrc0mmand** [comment]
Fuzzers are also still happy!

File: src/core/unit.c (hunk 2 above)
Comment (yuwata) [reply src/core/unit.c]: The comment in the code is updated. PTAL.

Comment (poettering) [reply src/core/unit.c]: but to find a suitable existing runtime object it doesn't really matter in which direction we go.

i.e. if a has JoinsNamespaceOf=b, and a starts first, it should allocate the namespace first, and b should then use it. And the exact same way the other way.

So why have this condition here? if it is dropped, what changes?

Comment (yuwata) [reply src/core/unit.c]: @poettering 
> but to find a suitable existing runtime object it doesn't really matter in which direction we go.
> 
> i.e. if a has JoinsNamespaceOf=b, and a starts first, it should allocate the namespace first, and b should then use it. And the exact same way the other way.

The current behavior does not work like that. If a.service has JoinsNamespaceOf=b.service, then a.service joins the namespaces of b.service if it is already started. But if a.service is started earlier, then b.service will not join the namespaces of a.service.

> So why have this condition here?

To keep the current behavior explained in the above. Previously, the reference introduced by JoinsNamespaceOf= is one-directional. But with the change, reference become bidirectional. Hence, we need to distinguish which unit introduce the dependency.

> if it is dropped, what changes?

Then, setting JoinsNamespaceOf=b.service in a.service implies JoinsNamespaceOf=a.service for b.service.

I think it is OK to 'fix' the current behavior. WDYT?

Comment (poettering) [reply src/core/unit.c]: > The current behavior does not work like that. If a.service has JoinsNamespaceOf=b.service, then a.service joins the namespaces of b.service if it is already started. But if a.service is started earlier, then b.service will not join the namespaces of a.service.

Uh, that's a bug! It was intended to be fully symmetric, i.e. A joins B should be entirely equivalent to B joins A... So if the condition is dropped this is actually a bugfix then.

**Comment by yuwata** [comment]
@poettering Fixed `JoinsNamespaceOf=`. PTAL.

File: src/core/unit.c [hunk 5]
```diff
@@ -6111,6 +6116,33 @@ int unit_get_dependency_array(const Unit *u, UnitDependencyAtom atom, Unit ***re
         return (int) n;
 }
 
+int unit_get_dependency_chain(Unit *u, UnitDependencyAtom atom, Set **ret) {
```
Comment (poettering) [inline src/core/unit.c]: is "chain" the right word? it's a whole subgraph after all, not just one path in it as the name "chain" would suggest.

maybe unit_get_transitive_dependency_set() or so?

i mean, that's what matters here, that we transitively pick up all deps.

**Comment by poettering** [comment]
lgtm, just wondering about naming of that function, but that's purely bike shedding

File: src/core/unit.c (hunk 5 above)
Comment (yuwata) [reply src/core/unit.c]: 'transitively'? I am not familiar with the word. I prefer 'chain' here. Or, how about unit_get_dependency_subgraph() or unit_get_graph_dependency_set() ? That is, 'chain' -> 'graph' or 'subgraph'?

Comment (poettering) [reply src/core/unit.c]: → https://en.wikipedia.org/wiki/Transitive_relation

it means that if we go from a to b and from b to c we can also go from a to c. which is what this is about here, the transitive completion of the graph.

Comment (poettering) [reply src/core/unit.c]: dependency_set() sounds ok too
subgraph() too

anything better than chain I think, since a chain suggests 1-dimenstional i think.

Comment (yuwata) [reply src/core/unit.c]: Ah, transitive relation! I did not know the name of the relation in English. Yeah, it sounds good.

File: src/core/unit.h [hunk 6]
```diff
@@ -810,6 +810,7 @@ static inline const UnitVTable* UNIT_VTABLE(const Unit *u) {
 
 Unit* unit_has_dependency(const Unit *u, UnitDependencyAtom atom, Unit *other);
 int unit_get_dependency_array(const Unit *u, UnitDependencyAtom atom, Unit ***ret_array);
+int unit_get_dependency_chain(Unit *u, UnitDependencyAtom atom, Set **ret);
```
Comment (YHNdnzj) [inline src/core/unit.h]: ../src/core/unit.c:6161:5: error: no previous prototype for function 'unit_get_transitive_dependency_set' [-Werror,-Wmissing-prototypes]
int unit_get_transitive_dependency_set(Unit *u, UnitDependencyAtom atom, Set **ret) {

Comment (yuwata) [reply src/core/unit.h]: Ouch.

**Comment by keszybz** [comment]
I didn't check the patches, because multiple people reviewed already. The test failure is unrelated.

**Pull request merged**